runtime/src/mips.rs

```diff
@@ -0,0 +1,721 @@
+        /// * `deposit` minimum deposit value
+        #[weight = SimpleDispatchInfo::FixedNormal(5_000_000)]
+        pub fn propose(origin, proposal: Box<T::Proposal>, deposit: BalanceOf<T>) -> Result {
+            let proposer = ensure_signed(origin)?;
+            let proposal_hash = T::Hashing::hash_of(&proposal);
+
+            // Pre conditions: caller must have min balance
+            ensure!(deposit >= Self::min_proposal_deposit(), "deposit is less than minimum required to start a proposal");
+            // Proposal must be new
+            ensure!(!<Proposals<T>>::exists(proposal_hash), "duplicate proposals are not allowed");
+
+            // Reserve the minimum deposit
+            T::Currency::reserve(&proposer, deposit).map_err(|_| "proposer can't afford to lock minimum deposit")?;
+
+            let index = Self::proposal_count();
+            <ProposalCount>::mutate(|i| *i += 1);
+
+            let proposal_meta = MipsMetadata {
+                index,
+                end: <system::Module<T>>::block_number() + Self::proposal_duration(),
+                proposal_hash
+            };
+            <ProposalMetadata<T>>::mutate(|metadata| metadata.push(proposal_meta));
+
+            <Deposits<T>>::insert(proposal_hash, (proposer.clone(), deposit));
+
+            <Proposals<T>>::insert(proposal_hash, *proposal);
+
+            let vote = Votes {
+                index,
+                ayes: vec![(proposer.clone(), deposit)],
+                nays: vec![],
+            };
+            <Voting<T>>::insert(proposal_hash, vote);
+
+            Self::deposit_event(RawEvent::Proposed(proposer, deposit, proposal_hash));
+            Ok(())
+        }
+
+        /// A network member can vote on any Mesh Improvement Proposal by selecting the hash that
+        /// corresponds ot the dispatchable action and vote with some balance.
+        ///
+        /// # Arguments
+        /// * `proposal` a dispatchable call
+        /// * `index` proposal index
+        /// * `aye_or_nay` a bool representing for or against vote
+        /// * `deposit` minimum deposit value
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn vote(origin, proposal_hash: T::Hash, index: MipsIndex, aye_or_nay: bool, deposit: BalanceOf<T>) {
+            let proposer = ensure_signed(origin)?;
+
+            let mut voting = Self::voting(&proposal_hash).ok_or("proposal does not exist")?;
+            ensure!(voting.index == index, "mismatched proposal index");
+
+            let position_yes = voting.ayes.iter().position(|(a, _)| a == &proposer);
+            let position_no = voting.nays.iter().position(|(a, _)| a == &proposer);
+
+            if position_yes.is_none() && position_no.is_none()  {
+                if aye_or_nay {
+                    voting.ayes.push((proposer.clone(), deposit));
+                } else {
+                    voting.nays.push((proposer.clone(), deposit));
+                }
```

Comment:
Do we not need to reserve the deposit to stop double-voting?

Reply:
Updated lofic to reserve deposit when a vote is cast. All deposits from proposer and voters become unreserved when proposal is closed. Added tests to make sure balances are reserved and unreserved.